versions/development/SPEC.md

```diff
@@ -1218,81 +1259,343 @@ $runtime = 'runtime' $ws* '{' ($ws* $runtime_kv $ws*)* '}'
+$hints_object = '{}' | '{' $hints_kv (, $hints_kv)* '}'
+$hints_array = '[]' |  '[' $hints_value (, $hints_value)* ']'
+```
+
+The hints section is purely optional and contains key/value pairs. The purpose of this section is to enable the author a means to encode information directly into the task which can be used by an execution engine, which understands the provided information, to optimize the execution of a task. An important distinction to make between `runtime` and `hints` is that a task cannot run without the parameters defined within the `runtime` section, however all properties within the `hints` section are optional and a task SHOULD be able to execute even if the engine completely ignored the entire hints section. Therefore, if there are specific  values within a `hints` section which an engine is expecting, then the engine must always define reasonable default values and not have a strict requirement for the author to provide them.
+
+It must also be clarified that the purpose of this section is **NOT** to provide an execution engine a means to encode all sorts of required values that it needs in order to execute a task without setting any default values. This practice should strictly be considered an anti-pattern and should be avoided by execution engines
+
+The values within the `hints` section are arbitrary K/V pairs, and it is generally up to the execution engine to define what values they are interested in. The provided values can be simple (`String`,`boolean`,`number`...)  or more complex (`Arrays`,`hints object`)
+
+```wdl
+task foo {
+  ...
+  hints {
+    some-key: "some-value"
+		
+  }
+}
+```
+
+#### Reserved Keys
+
+Although `hints` are arbitrary, there are a number of keys which are reserved by the language in or order to try and encourage interoperability of tasks and workflows between different execution engines. The list of Reserved keys is likely to grow, and in order to prevent name space collisions, engines should follow the [best practices](#conventions-and-best-practices) for specifying hints. 
+
+* **maxCpu**: Specify the maximum CPU to be provisioned for a task. It is up to the engine, whether or not it enforces the `maxCpu` hint. The type of this hint should be the same as `runtime.cpu`
+* **maxMemory**: Specify the maximum memory provisioned for a task. It is up to the engine, whether or not it enforces the `maxMemory` hint. the type of this hint should be the same as `runtime.memory`
+* **shortTask**: Tell the execution engine that this task is not expected to take long to execute, and therefore the engine can attempt to optimize the execution in whatever way it defines. This flag can be used to tell an engine to use the cost optimized instance types that many clouds provide (but are available only for a limited time). An example of this would be `preemptible` instances on `gcp` and `spot` instances on `aws`. "Short" is a bit relative, but should generally be interpreted as << 24h. 
```

Comment:
Is this preferable to something more granular? For example, this could instead be "duration" which is the expected duration of the task to some degree of specificity. The specificity could be categorical ("short", "medium", "long") or a continuous value ("48h").

Reply:
I actually like @jdidion's idea here. Some grids/clusters require you to set a maximum runtime of a job or choose a specific  queue based on how long a job takes. An engine could use a `duration` runtime attribute to make such choices in a more flexable manner then with a single flag specifying a task is "short".

Reply:
@DavyCats part of me wanted to avoid that on the first pass of this, mainly becuase thats a level of granularity that, although useful, might not be available when programatically running jobs with diverse  inputs. Additionally, it is not really applicable except in the case of Some grid/clusters. The runtime block is the "Required" section and I really dont want to have a required `minTime` and `maxTime` for EVERY execution on every platform since its only required in maybe 5% of situaations. Its harder to set sensible defaults for those as well. I would rather `minTime` and `maxTime` be part of the `hints` section then then the `runtime` section

Reply:
Chiming in later, but I agree that a `duration` / `maxTime` field would be less relative, and allow for better scheduling of jobs. For instance, today we were considering that if a job is less than 30 seconds, we might run it on the same instance as Cromwell to avoid spinning up or submitting a new job. For GCP, if the job is less than 24 hours you might consider submitting it as a preemptible job. 

You could always write an expression here to derive the duration based on the size or number of inputs etc.